### ISSUE TITLE
tsc with allowJs reports TS9005 error for nested anonymous constructor functions

### ISSUE THREAD

**Issue by brian6932** [microsoft/TypeScript#57523]
### 🔎 Search Terms

TS9005: Declaration emit for this file requires using private name '(Anonymous function)'. An explicit type annotation may unblock declaration emit.

### 🕗 Version & Regression Information

- Occurs on versions 3.7.5 - nightly
- Related to PR #55472

Related to #55172
When 2 separate nested `this` bindings are used within constructor functions, the following error's returned always at the top of the file (this was really annoying to debug 😅)
```
Declaration emit for this file requires using private name '(Anonymous function)'. An explicit type annotation may unblock declaration emit.
```

### ⏯ Playground Link

https://tsplay.dev/NV57Gw

### 💻 Code

```ts
const a = new function () {
	this.b = new function () {
		this.c = 1
	}
}
```

### 🙁 Actual behavior

Complains

### 🙂 Expected behavior

Shouldn't complain

**Comment by Andarist** [comment]
I'm not sure if this is meant to compile successfully but, if I'm reading this correctly, it shows an extra thing that is a bug. If you comment out the nested assignment to the `this.c` property then you might see this emit:
```ts
declare function a(): void;
declare class a {
    b: any;
}
```

This doesn't seem right because `a` is not a class - it's an instance of a class. Extra TS playground that shows the difference between anonymous and named functions here: [TS playground](https://www.typescriptlang.org/play?strict=false&noImplicitAny=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false&ts=5.5.0-dev.20240224&filetype=js&strictNullChecks=false#code/MYewdgzgLgBAhjAvDMBTA7jAFAMwK5jBQCW42AlDAN4BQAkFABbEQB0ADgE4jtIwCMABhoBfcjRr5CJMgCEslWg2ZsuPPkNE1QkWACM+aTLJpA)

And an extra thing that I observed here, private name for `#prop` can leak in the declaration emit - this shouldn't happen ([TS playground](https://www.typescriptlang.org/play?strict=false&noImplicitAny=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false&target=99&ts=5.5.0-dev.20240224&filetype=js&strictNullChecks=false#code/PTAEAEBcEMCcHMCmkBcpEGcB2iAekAoAMwFcsBjSASwHstRJNIAKASlAG8DRRZkTY9cgBtoGDKACCnbj1ABiAA6wai0AF5QARgAMAblkBfA4YIE8imrEihydDDegbQOAO6hmjB21ZsDQA)):
```ts
// input

// @target: esnext
function test() {
  return class A {
    #prop = 10;
  };
}

export const a = new (test())();

// dts
export declare const a: {
    "__#12@#prop": number;
};
```

**Comment by fatcerberus** [comment]
@Andarist Re: Private name leakage - https://github.com/microsoft/TypeScript/issues/56145 -> https://github.com/microsoft/TypeScript/issues/36548

**Comment by kopax** [comment]
Hello, any update on this ?

Using yarn4 and prisma, we get stuck as : 

- yarn4 can't generate prisma client in node_modules, so it is generated in the workspace with my files
- I can't use annotation as this is often re-generated code

This makes me totally blocked and forced to downgrade yarn :(. 

Any workaround exisiting ?

**Comment by Jeangowhy** [comment]
Can someone tell what is this problem?

**Comment by seahindeniz** [comment]
In my case, `"composite": true,` was in my tsconfig.json file. Removing it solved the issue
